cli/azd/pkg/infra/azure_resource_manager.go

```diff
@@ -67,33 +88,54 @@ func (rm *AzureResourceManager) GetDeploymentResourceOperations(
 	}
 
 	if strings.TrimSpace(resourceGroupName) == "" {
-		return resourceOperations, nil
+		return topLevelDeploymentOperations, nil
+	}
+
+	allLevelsDeploymentOperations := []*armresources.DeploymentOperation{}
+
+	resourceIdPrefix := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s", subscriptionId, resourceGroupName)
+	isWithinResourceGroup := func(operation *armresources.DeploymentOperation) bool {
+		return strings.HasPrefix(*operation.Properties.TargetResource.ID, resourceIdPrefix)
 	}
 
 	// Find all resource group deployments within the subscription operations
 	// Recursively append any resource group deployments that are found
-	deploymentOperations := make(map[string]*armresources.DeploymentOperation)
-	for _, operation := range resourceOperations {
-		if operation.Properties.TargetResource != nil &&
-			*operation.Properties.TargetResource.ResourceType == string(AzureResourceTypeDeployment) &&
-			*operation.Properties.ProvisioningOperation == armresources.ProvisioningOperationCreate {
+	innerLevelDeploymentOperations := make(map[string]*armresources.DeploymentOperation)
+	for _, operation := range topLevelDeploymentOperations {
+		if !hasTargetResourceData(operation) {
+			// Operations w/o target data can't be resolved. Ignoring them
+			continue
+		}
+		if !isWithinResourceGroup(operation) {
```

Comment:
Can you help me better understand this case?  What's a concrete example where this happens and why do we want to ignore the operation?

Reply:
While doing a deployment, some times, deployments are created outside of the resource group.
For example, for this APIM template: https://github.com/puicchan/app-templates-integration-services/blob/main/infra/main.bicep#L30

When we call `azd provision`, and we follow the link to the azure portal, there is a deployment listed there which is not part of the resource group (it is not even deploying anything, as it is a deployment type, but empty. With no other child resources).

After we start the deployment, we fetch all the top-level deployments (including any deployment which is not within the resource group), and then we hope to fetch information about the content of it by getting the deployment description from within the resource group.  But since it is not in the resource group, the operation fails.

Reply:
repro: `azd up -t https://github.com/puicchan/app-templates-integration-services`

List of resources from Azure portal deployment:

![image](https://user-images.githubusercontent.com/24213737/222821045-5e1cf420-ba7e-4139-b96c-2294c79f4367.png)

Note the deployment circled in red....

Then, list deployments from resource group:

![image](https://user-images.githubusercontent.com/24213737/222821179-708644fe-ba4d-474c-a100-69e4eafc3ad5.png)

Reply:
Thanks, this was very helpful for my understanding!

I think you might have intended to link to this part of the bicep: https://github.com/puicchan/app-templates-integration-services/blob/bac6c6f90c855e4345acd1f5d88070df3605dbc0/infra/main.bicep#L126-L141

I think I understand what was going on.  In this sample, the deployment we are doing at the subscription level for `main.bicep` ends up doing another subscription level deployment, and that was confusing our logic, because we assumed that any deployment, we discovered from the subscription level was going to be a deployment that targeted a resource group (since that's what most of our templates do, they create a resource group using the subscription level deployment and then do a nested deployment into that resource group to deploy the resources).

The goal here is to avoid recuring into these subscription level deployments as we try to collect operations, because our recuring code is built on the assumption that the deployments always target resource groups.

This seems like a fine tactical fix for now.  I think the right thing long term is that we examine the deployment to understand the scope it targets and then takes that into account when discovering all the resource operations.  I'm a little worried that this strategy is not yet fully complete and there may be cases where don't pull a complete set of operations.

Some interesting cases for us to look at:

1. A deployment at subscription scope which creates two resource groups and does a deployment into each RG.
2. A deployment at subscription scope which creates another deployment at subscription scope which creates a resource group and then does a deploy into it.

I think in the fullness of time, GetDeploymentResourceOperations should be able to collect all the operations in both of these cases, and my gut says right now we might still have a few issues here.